src/solver.hs

```diff
@@ -0,0 +1,232 @@
+                                                        pos (Note [Candidate v1, Candidate v2])
+                                        _ -> error "Not a naked pair"
+
+-- | Returns the value of a single position in a row
+valueFromSPR :: Sudoku -> Position -> Value
+valueFromSPR sud pos = valueFromSPSection sud pos (rowPositions pos)
+
+-- | Returns the value of a single position in a column
+valueFromSPC :: Sudoku -> Position -> Value
+valueFromSPC sud pos = valueFromSPSection sud pos (colPositions pos)
+
+-- | Returns the value of a single position in a block
+valueFromSPB :: Sudoku -> Position -> Value
+valueFromSPB sud pos = valueFromSPSection sud pos (blockPositions pos)
+
+-- | Returns the value of a single position in a section
+valueFromSPSection :: Sudoku -> Position -> [Position] -> Value
+valueFromSPSection sud pos secPos = case getSinglePosition sud pos secPos of
+                                        [x] -> Filled x
+                                        _ -> error "Not a single position"
+
+-- | Returns the value of a single candidate in a cell
+valueFromSC :: Sudoku -> Position -> Value
+valueFromSC sud pos = case candidates of
+                        [x] -> Filled x
+                        _ -> error "Not a single candidate"
+    where
+        candidates = getCandidates sud pos
+
+-- | Last free cell row
+valueFromLFCR :: Sudoku -> Position -> Value
+valueFromLFCR sud pos = valueFromLFCSection (rowFromPos sud pos)
+
+-- | Last free cell collumn
+valueFromLFCC :: Sudoku -> Position -> Value
+valueFromLFCC sud pos = valueFromLFCSection (colFromPos sud pos)
+
+-- | Last free cell block
+valueFromLFCB :: Sudoku -> Position -> Value
+valueFromLFCB sud pos = valueFromLFCSection (blockFromPos sud pos)
+
+-- | Returns the value of the last free cell in a section
+valueFromLFCSection :: Section -> Value
+valueFromLFCSection sec = case list' of
+                            [x] -> x
+                            _ -> error "Not a Section does not work with lemma LFCS"
+    where
+        list' = [Filled x | x <- [One ..]] \\ sec
+
+-- | Returns the value of a candidate line
+valueFromCL :: Sudoku -> Position -> Value
+valueFromCL sud pos = case filter (\(l,v) -> l /= []) lines of
+                        as -> Note $ map (\([l],v) -> Line l v) as
+                        -- _ -> error "Not a single candidate" -- Overlapping pattern, might need to change this if it is ia problem
+    where
+        candidates = getCandidates sud pos
+        lines = [(lineBlock sud pos l, l) | l <- candidates, lineBlock sud pos l /= []]
+
+-- | next step in solving the sudoku
+nextStep :: Sudoku -> [Position -> Step] -> Int -> (Step,Int)
+nextStep sud [] count = (NOAVAILABLESTEPS,count)
+nextStep sud (sf:sfs) count = case fst res of
+                            NOAVAILABLESTEPS -> nextStep sud sfs (count + snd res)
+                            x -> res
+        where
+                res = tryStepsOnEmpty sud sf count
+
```

Comment:
Could use a [monad](https://wiki.haskell.org/Monad) to encapsulate the counting so that you don't have to manually pass it around:
```haskell
State Int x = Int -> (x, Int) 
```
This function would return `State Int Step`.

Reply:
(Advanced: use [traverse](https://en.wikibooks.org/wiki/Haskell/Traversable).)